



<!DOCTYPE html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="The geneXplain platform Java API manual">
      
      
      
        <meta name="author" content="pstegma">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-3.0.6">
    
    
      
        <title>JAR application manuals - The geneXplain Java API documentation</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/application.451f80e5.css">
      
      
    
    
      <script src="../assets/javascripts/modernizr.1aa3b519.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700|Roboto+Mono">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="../assets/fonts/material-icons.css">
    
    
  </head>
  
    <body dir="ltr">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448"
    viewBox="0 0 416 448" id="__github">
  <path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19-18.125
        8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19 18.125-8.5
        18.125 8.5 10.75 19 3.125 20.5zM320 304q0 10-3.125 20.5t-10.75
        19-18.125 8.5-18.125-8.5-10.75-19-3.125-20.5 3.125-20.5 10.75-19
        18.125-8.5 18.125 8.5 10.75 19 3.125 20.5zM360
        304q0-30-17.25-51t-46.75-21q-10.25 0-48.75 5.25-17.75 2.75-39.25
        2.75t-39.25-2.75q-38-5.25-48.75-5.25-29.5 0-46.75 21t-17.25 51q0 22 8
        38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0
        37.25-1.75t35-7.375 30.5-15 20.25-25.75 8-38.375zM416 260q0 51.75-15.25
        82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5-41.75
        1.125q-19.5 0-35.5-0.75t-36.875-3.125-38.125-7.5-34.25-12.875-30.25-20.25-21.5-28.75q-15.5-30.75-15.5-82.75
        0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25
        30.875q36.75-8.75 77.25-8.75 37 0 70 8 26.25-20.5
        46.75-30.25t47.25-9.75q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34
        99.5z" />
</svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#jar-application-manuals" tabindex="1" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href=".." title="The geneXplain Java API documentation" class="md-header-nav__button md-logo">
          
            <i class="md-icon"></i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            
              <span class="md-header-nav__topic">
                The geneXplain Java API documentation
              </span>
              <span class="md-header-nav__topic">
                JAR application manuals
              </span>
            
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          
            <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
            
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
          
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  


  <a href="https://github.com/genexplain/genexplain-api/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href=".." title="The geneXplain Java API documentation" class="md-nav__button md-logo">
      
        <i class="md-icon"></i>
      
    </a>
    The geneXplain Java API documentation
  </label>
  
    <div class="md-nav__source">
      


  


  <a href="https://github.com/genexplain/genexplain-api/" title="Go to repository" class="md-source" data-md-source="github">
    
      <div class="md-source__icon">
        <svg viewBox="0 0 24 24" width="24" height="24">
          <use xlink:href="#__github" width="24" height="24"></use>
        </svg>
      </div>
    
    <div class="md-source__repository">
      GitHub
    </div>
  </a>

    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      


  <li class="md-nav__item">
    <a href=".." title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="../java_api/" title="Java API" class="md-nav__link">
      Java API
    </a>
  </li>

    
      
      
      

  


  <li class="md-nav__item md-nav__item--active md-nav__item--nested">
    
      <input class="md-toggle md-nav__toggle" data-md-toggle="nav-3" type="checkbox" id="nav-3" checked>
    
    <label class="md-nav__link" for="nav-3">
      JAR applications and JSON interface
    </label>
    <nav class="md-nav" data-md-component="collapsible" data-md-level="1">
      <label class="md-nav__title" for="nav-3">
        JAR applications and JSON interface
      </label>
      <ul class="md-nav__list" data-md-scrollfix>
        
        
          
          
          


  <li class="md-nav__item">
    <a href="../java_json_manual_intro/" title="Introduction" class="md-nav__link">
      Introduction
    </a>
  </li>

        
          
          
          

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        JAR application manuals
      </label>
    
    <a href="./" title="JAR application manuals" class="md-nav__link md-nav__link--active">
      JAR application manuals
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example-example-applications-using-the-java-api" title="example - Example applications using the Java API" class="md-nav__link">
    example - Example applications using the Java API
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#apps-listing-available-tools" title="apps - Listing available tools" class="md-nav__link">
    apps - Listing available tools
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regulator-search-regulator-and-effector-search-tool" title="regulator-search - Regulator (and effector) search tool" class="md-nav__link">
    regulator-search - Regulator (and effector) search tool
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters" title="Parameters" class="md-nav__link">
    Parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-json-config-for-the-java-application" title="Example JSON config for the Java application" class="md-nav__link">
    Example JSON config for the Java application
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-json-output-of-the-regulatoreffector-search-application" title="Example JSON output of the Regulator/Effector search application" class="md-nav__link">
    Example JSON output of the Regulator/Effector search application
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exec-executing-tasks-using-the-json-interface" title="exec - Executing tasks using the JSON interface" class="md-nav__link">
    exec - Executing tasks using the JSON interface
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#main-executors" title="Main executors" class="md-nav__link">
    Main executors
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#analyze" title="analyze" class="md-nav__link">
    analyze
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#branch" title="branch" class="md-nav__link">
    branch
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#createfolder" title="createFolder" class="md-nav__link">
    createFolder
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#export" title="export" class="md-nav__link">
    export
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#external" title="external" class="md-nav__link">
    external
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get" title="get" class="md-nav__link">
    get
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#import" title="imPort" class="md-nav__link">
    imPort
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#itemparameters" title="itemParameters" class="md-nav__link">
    itemParameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#jobstatus" title="jobStatus" class="md-nav__link">
    jobStatus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#listitems" title="listItems" class="md-nav__link">
    listItems
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#list" title="list" class="md-nav__link">
    list
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#put" title="put" class="md-nav__link">
    put
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#setparameters" title="setParameters" class="md-nav__link">
    setParameters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

        
          
          
          


  <li class="md-nav__item">
    <a href="../json_hello_world/" title="JSON interface Hello world tutorial" class="md-nav__link">
      JSON interface Hello world tutorial
    </a>
  </li>

        
      </ul>
    </nav>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#example-example-applications-using-the-java-api" title="example - Example applications using the Java API" class="md-nav__link">
    example - Example applications using the Java API
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#apps-listing-available-tools" title="apps - Listing available tools" class="md-nav__link">
    apps - Listing available tools
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#regulator-search-regulator-and-effector-search-tool" title="regulator-search - Regulator (and effector) search tool" class="md-nav__link">
    regulator-search - Regulator (and effector) search tool
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#parameters" title="Parameters" class="md-nav__link">
    Parameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-json-config-for-the-java-application" title="Example JSON config for the Java application" class="md-nav__link">
    Example JSON config for the Java application
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#example-json-output-of-the-regulatoreffector-search-application" title="Example JSON output of the Regulator/Effector search application" class="md-nav__link">
    Example JSON output of the Regulator/Effector search application
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#exec-executing-tasks-using-the-json-interface" title="exec - Executing tasks using the JSON interface" class="md-nav__link">
    exec - Executing tasks using the JSON interface
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#main-executors" title="Main executors" class="md-nav__link">
    Main executors
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#analyze" title="analyze" class="md-nav__link">
    analyze
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#branch" title="branch" class="md-nav__link">
    branch
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#createfolder" title="createFolder" class="md-nav__link">
    createFolder
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#export" title="export" class="md-nav__link">
    export
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#external" title="external" class="md-nav__link">
    external
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#get" title="get" class="md-nav__link">
    get
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#import" title="imPort" class="md-nav__link">
    imPort
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#itemparameters" title="itemParameters" class="md-nav__link">
    itemParameters
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#jobstatus" title="jobStatus" class="md-nav__link">
    jobStatus
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#listitems" title="listItems" class="md-nav__link">
    listItems
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#list" title="list" class="md-nav__link">
    list
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#put" title="put" class="md-nav__link">
    put
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#setparameters" title="setParameters" class="md-nav__link">
    setParameters
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/genexplain/genexplain-api/edit/master/docs/json_interface.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="jar-application-manuals">JAR application manuals</h1>
<h2 id="example-example-applications-using-the-java-api">example - Example applications using the Java API</h2>
<p>Several examples show how to use the Java API. They can be listed by invoking the JAR with the <em>example</em> command. All of the example sources can be found in the package <em>com.genexplain.api.eg</em>.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>genexplain-api$ java -jar build/libs/genexplain-api-1.0.jar example
INFO  com.genexplain.api.app.APIRunner - Running <span class="nb">command</span> example
INFO  com.genexplain.api.eg.ExampleRunner - createFolder - Creates a folder within the demo workspace and prints out the server response <span class="o">(</span>com.genexplain.api.eg.CreateFolderExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - extractGeneClasses - Extracts gene classes from a functional classification result applying criteria <span class="k">for</span> filtering <span class="o">(</span>com.genexplain.api.eg.ExtractFunctionalClassGenesExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - getParameters - Fetches parameter descriptions <span class="k">for</span> a specified analysis tool <span class="o">(</span>com.genexplain.api.eg.GetParametersExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - getTable - Gets a JSON response representing a data table and prints it to standard output <span class="o">(</span>com.genexplain.api.eg.GetTableExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - importTable - Imports a table to the specified path <span class="o">(</span>com.genexplain.api.eg.ImportTableExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - listAFolder - Gets a JSON response representing the contents of the data/Projects folder <span class="o">(</span>com.genexplain.api.eg.ListAFolderExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - putTable - Stores a table under the specified path <span class="o">(</span>com.genexplain.api.eg.PutTableExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - tfbsAnalysisForFolder - Analyzes binding site enrichment <span class="k">for</span> all gene sets in a folder <span class="o">(</span>com.genexplain.api.eg.TfbsAnalysisForFolderExample<span class="o">)</span>
INFO  com.genexplain.api.eg.ExampleRunner - zipImport - Imports multiple files of same <span class="nb">type</span> as a ZIP archive <span class="o">(</span>com.genexplain.api.eg.ZipImportExample<span class="o">)</span>
</pre></div>
</td></tr></table>

<p>Each listed item represents an example that can be invoked by adding its name to the commandline, e.g. <em>listAFolder</em>:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>genexplain-api$ java -jar build/libs/genexplain-api-1.0.jar example listAFolder
INFO  com.genexplain.api.app.APIRunner - Running <span class="nb">command</span> example                                                                                                                                                 
INFO  c.g.api.core.GxHttpConnectionImpl - Trying to log in                                                                                                                                                       
INFO  c.g.api.eg.ListAFolderExample - <span class="o">{</span><span class="s2">&quot;names&quot;</span>:<span class="o">[{</span><span class="s2">&quot;hasChildren&quot;</span>:true,<span class="s2">&quot;permissions&quot;</span>:3,<span class="s2">&quot;name&quot;</span>:<span class="s2">&quot;CRC_6_CpG_biomarkers&quot;</span>,<span class="s2">&quot;protection&quot;</span>:2,<span class="s2">&quot;class&quot;</span>:0<span class="o">}</span>,<span class="o">{</span><span class="s2">&quot;hasChildren&quot;</span>:true,<span class="s2">&quot;permissions&quot;</span>:7,<span class="s2">&quot;name&quot;</span>:<span class="s2">&quot;Demo project&quot;</span>,<span class="s2">&quot;protection&quot;</span>:2,<span class="s2">&quot;class&quot;</span>:0<span class="o">}</span>,<span class="o">{</span><span class="s2">&quot;hasChildren&quot;</span>:true,<span class="s2">&quot;permissions&quot;</span>:3,<span class="s2">&quot;name&quot;</span>:<span class="s2">&quot;MTX resistance&quot;</span>,<span class="s2">&quot;protection&quot;</span>:2,<span class="s2">&quot;class&quot;</span>:0<span class="o">}]</span>,<span class="s2">&quot;size&quot;</span>:3,<span class="s2">&quot;classes&quot;</span>:<span class="o">[</span><span class="s2">&quot;biouml.model.Module&quot;</span><span class="o">]</span>,<span class="s2">&quot;from&quot;</span>:0,<span class="s2">&quot;to&quot;</span>:3,<span class="s2">&quot;enabled&quot;</span>:true<span class="o">}</span>
</pre></div>
</td></tr></table>

<p>The example applications <em>extractGeneClasses</em>, <em>tfbsAnalysisForFolder</em> and <em>zipImport</em> require a configuration file in JSON format as additional commandline argument. Please refer to their class source codes (<code>com.genexplain.api.eg.ExtractFunctionalClassGenesExample</code>,  <code>com.genexplain.api.eg.TfbsAnalysisForFolderExample</code>, <code>com.genexplain.api.eg.ZipImportExample</code>) for descriptions of the configurable parameters. Runnable example JSON files are provided in the <em>docs/tutorial</em> folder of the source repository. They are also shown below. Please note that some parts have been abbreviated.</p>
<p>The <em>extractGeneClasses</em> demo extracts gene sets from the functional classification result specified by <code>funClassTable</code> firstly to a local folder named <em>fun_class_gene_sets</em> (specified by <code>geneSetFolder</code>) and imports them to an output folder (<code>outputFolder</code>) in <em>data/Projects/Demo project/Data</em> (<code>geneSetPath</code>) using the default result folder name (<em>fun_class_gene_sets</em>).</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;funClassTable&quot;</span><span class="p">:</span>  <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2...&quot;</span><span class="p">,</span>
    <span class="nt">&quot;localTableFile&quot;</span><span class="p">:</span> <span class="s2">&quot;funclass_table.txt&quot;</span><span class="p">,</span>
    <span class="nt">&quot;minGroupSize&quot;</span><span class="p">:</span> <span class="mi">50</span><span class="p">,</span>
    <span class="nt">&quot;maxGroupSize&quot;</span><span class="p">:</span> <span class="mi">1500</span><span class="p">,</span>
    <span class="nt">&quot;minHits&quot;</span><span class="p">:</span> <span class="mi">30</span><span class="p">,</span>
    <span class="nt">&quot;maxHits&quot;</span><span class="p">:</span> <span class="mi">150</span><span class="p">,</span>
    <span class="nt">&quot;maxAdjustedPval&quot;</span><span class="p">:</span> <span class="mf">1e-10</span><span class="p">,</span>
    <span class="nt">&quot;geneSetFolder&quot;</span><span class="p">:</span> <span class="s2">&quot;fun_class_gene_sets&quot;</span><span class="p">,</span>
    <span class="nt">&quot;geneSetPath&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Projects/Demo project/Data&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<p>The <em>tfbsAnalysisForFolder</em> demo conducts a <a href="https://platform.genexplain.com/bioumlweb/#de=analyses/Methods/Site%20analysis/Search%20for%20enriched%20TFBSs%20(genes)">search for enriched transcription factor binding sites</a> for all gene sets of a folder (<code>inputFolder</code>). Gene sets of interest can be recognized by a regular expression (<code>inputRegex</code>), which defaults to <code>.+</code> if omitted from the configuration. The gene set for comparison (<code>noSet</code>) is a table with Ensembl gene ids which shouldn't overlap with the gene sets of interest.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;inputFolder&quot;</span><span class="p">:</span>    <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2...&quot;</span><span class="p">,</span>
    <span class="nt">&quot;inputRegex&quot;</span><span class="p">:</span>     <span class="s2">&quot;.+response.+&quot;</span><span class="p">,</span>
    <span class="nt">&quot;outputFolder&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Projects/Demo project/Data/TFBS analysis for folder&quot;</span><span class="p">,</span>
    <span class="nt">&quot;noSet&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2...&quot;</span><span class="p">,</span>
    <span class="nt">&quot;profile&quot;</span><span class="p">:</span> <span class="s2">&quot;databases/TRANSFAC(R) (public)/Data/profiles/vertebrate_human_p0.001&quot;</span><span class="p">,</span>
    <span class="nt">&quot;doSample&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;sampleNum&quot;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
    <span class="nt">&quot;sampleSize&quot;</span><span class="p">:</span> <span class="mi">1000</span><span class="p">,</span>
    <span class="nt">&quot;from&quot;</span><span class="p">:</span> <span class="mi">-1000</span><span class="p">,</span>
    <span class="nt">&quot;to&quot;</span><span class="p">:</span> <span class="mi">100</span><span class="p">,</span>
    <span class="nt">&quot;siteFeCutoff&quot;</span><span class="p">:</span> <span class="mf">1.1</span><span class="p">,</span>
    <span class="nt">&quot;siteFdrCutoff&quot;</span><span class="p">:</span> <span class="mf">1e-4</span><span class="p">,</span>
    <span class="nt">&quot;seqFeCutoff&quot;</span><span class="p">:</span> <span class="mf">1.0</span><span class="p">,</span>
    <span class="nt">&quot;seqFdrCutoff&quot;</span><span class="p">:</span> <span class="mf">0.05</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<p>The demo application <em>zipImport</em> imports a ZIP archive containing multiple files of the same type into the platform workspace. Besides login credentials, the local path of the archive and the platform destination folder one needs to specify the parameters for the file type-specific importer. Importer parameters for two frequent use cases, gene tables and CEL files, are also given in the source code. Parameters of the JSON shown below may require adaptation.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;user&quot;</span><span class="p">:</span> <span class="s2">&quot;optional user name&quot;</span><span class="p">,</span>
    <span class="nt">&quot;password&quot;</span><span class="p">:</span> <span class="s2">&quot;optional password&quot;</span><span class="p">,</span>
    <span class="nt">&quot;server&quot;</span><span class="p">:</span> <span class="s2">&quot;optional server URL&quot;</span><span class="p">,</span>
    <span class="nt">&quot;zipArchive&quot;</span><span class="p">:</span> <span class="s2">&quot;local_example_archive.zip&quot;</span><span class="p">,</span>
    <span class="nt">&quot;outputFolder&quot;</span><span class="p">:</span> <span class="s2">&quot;destination folder in platform&quot;</span><span class="p">,</span>
    <span class="nt">&quot;importParams&quot;</span><span class="p">:</span> <span class="p">[</span>
          <span class="p">{</span><span class="nt">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;cleanupFolder&quot;</span><span class="p">,</span><span class="nt">&quot;value&quot;</span><span class="p">:</span><span class="kc">false</span><span class="p">},</span>
          <span class="p">{</span><span class="nt">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;preserveExtension&quot;</span><span class="p">,</span><span class="nt">&quot;value&quot;</span><span class="p">:</span><span class="kc">false</span><span class="p">},</span>
          <span class="p">{</span><span class="nt">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;preserveArchiveStructure&quot;</span><span class="p">,</span><span class="nt">&quot;value&quot;</span><span class="p">:</span><span class="kc">false</span><span class="p">},</span>
          <span class="p">{</span><span class="nt">&quot;name&quot;</span><span class="p">:</span><span class="s2">&quot;importFormat&quot;</span><span class="p">,</span><span class="nt">&quot;value&quot;</span><span class="p">:</span><span class="s2">&quot;Affymetrix CEL file (*.cel)&quot;</span><span class="p">}</span>
    <span class="p">]</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h2 id="apps-listing-available-tools">apps - Listing available tools</h2>
<p>The application named <em>apps</em> produces a listing of the available analysis tools on a certain server. It takes a single argument that specifies
a file containing a JSON object with several properties of which only the <em>server</em> property is required. An example input file can be found in the <em>json</em> folder of this repository. Invoking the application with this file lists platform tools available for the <em>demo</em> account on the
<em>platform.genexplain.com</em> server as shown here:</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19</pre></div></td><td class="code"><div class="codehilite"><pre><span></span>genexplain-api$ java -jar build/libs/genexplain-api-1.0.jar apps json/application_lister_demo.json 
INFO  com.genexplain.api.app.APIRunner - Running <span class="nb">command</span> apps
Data manipulation/Annotate diagram
Data manipulation/Annotate table
Data manipulation/Annotate track with genes
Data manipulation/Composite module to proteins
Data manipulation/Convert table
Data manipulation/Convert table to track
Data manipulation/Convert table via homology
Data manipulation/Create folder
Data manipulation/Create miRNA promoters
Data manipulation/Create random track
Data manipulation/Create tissue-specific promoter track
Data manipulation/Create transcript region track
Data manipulation/Filter one track by another
Data manipulation/Filter table
Data manipulation/Filter track by condition
Data manipulation/Gene <span class="nb">set</span> to track
<span class="o">[</span>...<span class="o">]</span>
</pre></div>
</td></tr></table>

<p>The parameters that can be specified with the JSON input file are described in the following table.</p>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>server</strong></td>
<td>Server URL to connect to</td>
</tr>
<tr>
<td><strong>user</strong></td>
<td>Username to use for connection</td>
</tr>
<tr>
<td><strong>password</strong></td>
<td>Password that belongs to the username</td>
</tr>
<tr>
<td><strong>withParameters</strong></td>
<td>If <code>true</code> will also list the platform tool parameters</td>
</tr>
<tr>
<td><strong>connection</strong></td>
<td>Package and name to locate a Java class to use for connection. This class must implement <em>com.genexplain.api.core.GxHttpConnection</em>.</td>
</tr>
<tr>
<td><strong>client</strong></td>
<td>Package and name to locate a Java class to use as platform client. This class must implement <em>com.genexplain.api.core.GxHttpClient</em>.</td>
</tr>
</tbody>
</table>
<h2 id="regulator-search-regulator-and-effector-search-tool">regulator-search - Regulator (and effector) search tool</h2>
<p>The <em>regulator-search</em> is a molecular network analysis tool that searches for important signaling molecules that may regulate the activity of input molecules. One can also search for common effectors of input molecules, which is called effector search. While this application is just named after the search for regulators, it offers both options. Here we introduce the overall concept, input data and parameters. Research articles describe the algorithm in more detail.</p>
<p>The basis for this network analysis is formed by a database which describes the molecular network consisting of molecules and molecular complexes (nodes) and their interactions as well as biochemical reactions (edges). Such a network is illustrated in the following figure.</p>
<p><img alt="Figure" src="../figures/molecular_network_300.png" /></p>
<p>A set of input molecules is located within the network (blue nodes), and many of them are connected via one or more (directed) interaction/reaction steps. In the regulator search, those nodes from which <em>outgoing</em> reaction cascades can reach one or more input molecules are potential regulators, whereas in the effector search the cascades start from the input molecules and are <em>incoming</em> from the point of a potential effector.</p>
<p>The algorithm searches for target (effector or regulator) nodes within a specified maximal number of interaction/reaction steps. An illustration of such a search path connecting input molecules and an upstream regulator is shown in the following figure.</p>
<p><img alt="Figure" src="../figures/network_search_path_300.png" /></p>
<p>To prioritize potential target nodes, the search routine calculates a score for each of them. The score takes into account several parameters such as the number of hits in the input set or the number of hits outside of the input set within the same search distance. Optionally, one can also specify weights for the input molecules as well as so-called <em>decorations</em> of the molecular network which can be used to associate weights with all network molecules, to add new interactions/reactions between molecules, or to omit nodes from the network as depicted in the following figure.</p>
<p><img alt="Figure" src="../figures/decorated_network_300.png" />.</p>
<p>To obtain a normalized score, one can choose to run the algorithm for randomized input sets with other parameters fixed  and to calculate a Z-score as well as a FDR.</p>
<p>This application can either invoke the analysis on a specified platform server or just print out a corresponding JSON configuration. The printed JSON can be used with the <em>exec</em> command e.g. as part of a pipeline.</p>
<h3 id="parameters">Parameters</h3>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>searchEffectors</strong></td>
<td>Set <code>true</code> to do effector (downstream from inputs) instead of regulator (upstream from inputs) search</td>
</tr>
<tr>
<td><strong>server</strong></td>
<td>Server URL to connect to</td>
</tr>
<tr>
<td><strong>user</strong></td>
<td>Username to use for connection</td>
</tr>
<tr>
<td><strong>password</strong></td>
<td>Password that belongs to the username</td>
</tr>
<tr>
<td><strong>verbose</strong></td>
<td>Set <code>true</code> get more progress info</td>
</tr>
<tr>
<td><strong>reconnect</strong></td>
<td>Set <code>true</code> to allow attempts to reconnect in case the connection is interrupted</td>
</tr>
<tr>
<td><strong>connection-class</strong></td>
<td>Package and name of a Java class that extends com.genexplain.api.core.GxHttpConnection and will be used instead of the standard class</td>
</tr>
<tr>
<td><strong>client-class</strong></td>
<td>Package and name of a Java class that extends com.genexplain.api.core.GxHttpClient and will be used instead of the standard class</td>
</tr>
<tr>
<td><strong>replaceStrings</strong></td>
<td>Define string labels to be replaced by specified input parameters</td>
</tr>
<tr>
<td><strong>sourcePath</strong></td>
<td>Platform path of the input molecule set</td>
</tr>
<tr>
<td><strong>weightColumn</strong></td>
<td>Name of column in input table to assign weights to molecules (optional)</td>
</tr>
<tr>
<td><strong>isInputSizeLimited</strong></td>
<td>Set <code>true</code> to apply the input size limit</td>
</tr>
<tr>
<td><strong>inputSizeLimit</strong></td>
<td>Use only the first specified number of input molecules</td>
</tr>
<tr>
<td><strong>maxRadius</strong></td>
<td>Max. number of reaction steps that may connect input molecules and regulator/effector</td>
</tr>
<tr>
<td><strong>scoreCutoff</strong></td>
<td>Score cutoff to filter regulators</td>
</tr>
<tr>
<td><strong>bioHub</strong></td>
<td>Database with molecular network that are applied in the analysis</td>
</tr>
<tr>
<td><strong>species</strong></td>
<td>Species of input molecules</td>
</tr>
<tr>
<td><strong>calculatingFDR</strong></td>
<td>Set <code>true</code> to calculate FDR for regulators</td>
</tr>
<tr>
<td><strong>FDRcutoff</strong></td>
<td>FDR cutoff to filter regulators</td>
</tr>
<tr>
<td><strong>ZScoreCutoff</strong></td>
<td>Z-score cutoff to filter regulators</td>
</tr>
<tr>
<td><strong>penalty</strong></td>
<td>Penalty factor for false positives, that is molecules connected to a regulator within <code>maxRadius</code> steps that are not input molecules</td>
</tr>
<tr>
<td><strong>contextDecorators</strong></td>
<td>Augment network component with context values such as node weights</td>
</tr>
<tr>
<td><strong>tableName</strong></td>
<td>Parameter for <code>contextDecorators</code>: Platform path of a table with context nodes and weights</td>
</tr>
<tr>
<td><strong>tableColumn</strong></td>
<td>Parameter for <code>contextDecorators</code>: Column in corresponding table <code>tableName</code> that contains weights</td>
</tr>
<tr>
<td><strong>decayFactor</strong></td>
<td>Factor that specifies decay of context weights</td>
</tr>
<tr>
<td><strong>removeNodeDecorators</strong></td>
<td>Specify nodes to omit from the network. These are not considered in the analysis.</td>
</tr>
<tr>
<td><strong>inputTable</strong></td>
<td>Parameter for <code>removeNodeDecorators</code>: Platform path of a table with molecules to remove</td>
</tr>
<tr>
<td><strong>isoformFactor</strong></td>
<td>Set <code>true</code> to normalize weights of multi-forms</td>
</tr>
<tr>
<td><strong>outputTable</strong></td>
<td>Platform path for the result table</td>
</tr>
<tr>
<td><strong>justPrint</strong></td>
<td>Set <code>true</code> to just print the parameters that would be sent to the platform</td>
</tr>
<tr>
<td><strong>wait</strong></td>
<td>Set <code>true</code> to wait for analysis to complete instead of starting the job asynchronously</td>
</tr>
<tr>
<td><strong>progress</strong></td>
<td>Set <code>true</code> to obtain progress information</td>
</tr>
</tbody>
</table>
<h3 id="example-json-config-for-the-java-application">Example JSON config for the Java application</h3>
<p>This is a runnable JSON example for the demo account that can be provided to the <em>regulator-search</em> command within an input file, e.g. by calling something like
<code>java -jar genexplain-api.jar regulator-search example.json</code>.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;searchEffectors&quot;</span><span class="p">:</span> <span class="kc">false</span><span class="p">,</span>
    <span class="nt">&quot;user&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="nt">&quot;password&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="nt">&quot;server&quot;</span><span class="p">:</span> <span class="s2">&quot;https://platform.genexplain.com&quot;</span><span class="p">,</span>
    <span class="nt">&quot;verbose&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;reconnect&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;sourcePath&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2639, Affymetrix HG-U133A microarray/Data/DEGs with limma/Normalized (RMA) DEGs with limma/Condition_1 vs. Condition_2/Up-regulated genes Ensembl_Enriched_motifs/Enriched motifs Summary filtered TFs Genes Entrez&quot;</span><span class="p">,</span>
    <span class="nt">&quot;weightColumn&quot;</span><span class="p">:</span> <span class="s2">&quot;Avg. adj. site FE&quot;</span><span class="p">,</span>
    <span class="nt">&quot;isInputSizeLimited&quot;</span><span class="p">:</span> <span class="kc">false</span><span class="p">,</span>
    <span class="nt">&quot;inputSizeLimit&quot;</span><span class="p">:</span> <span class="mi">100000</span><span class="p">,</span>
    <span class="nt">&quot;maxRadius&quot;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
    <span class="nt">&quot;scoreCutoff&quot;</span><span class="p">:</span> <span class="mf">0.1</span><span class="p">,</span>
    <span class="nt">&quot;bioHub&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneWays hub&quot;</span><span class="p">,</span>
    <span class="nt">&quot;species&quot;</span><span class="p">:</span> <span class="s2">&quot;Human (Homo sapiens)&quot;</span><span class="p">,</span>
    <span class="nt">&quot;calculatingFDR&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;FDRcutoff&quot;</span><span class="p">:</span> <span class="mf">0.01</span><span class="p">,</span>
    <span class="nt">&quot;ZScoreCutoff&quot;</span><span class="p">:</span> <span class="mf">1.1</span><span class="p">,</span>
    <span class="nt">&quot;penalty&quot;</span><span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
    <span class="nt">&quot;isoformFactor&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;outputTable&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Projects/Demo project/example_regulator_search&quot;</span><span class="p">,</span>
    <span class="nt">&quot;contextDecorators&quot;</span><span class="p">:</span> <span class="p">[</span>
        <span class="p">{</span>
            <span class="nt">&quot;tableName&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2639, Affymetrix HG-U133A microarray/Data/DEGs with limma/Treatment vs. Control Genes Entrez&quot;</span><span class="p">,</span>
            <span class="nt">&quot;tableColumn&quot;</span><span class="p">:</span> <span class="s2">&quot;logFC&quot;</span><span class="p">,</span>
            <span class="nt">&quot;decayFactor&quot;</span><span class="p">:</span> <span class="mf">0.01</span>
        <span class="p">}</span>
    <span class="p">],</span>
    <span class="nt">&quot;wait&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;progress&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;justPrint&quot;</span><span class="p">:</span> <span class="kc">true</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="example-json-output-of-the-regulatoreffector-search-application">Example JSON output of the Regulator/Effector search application</h3>
<p>This is example JSON output that is generated for the above example configuration. This output can be used with the JSON executor using the <em>exec</em> command.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre> 1
 2
 3
 4
 5
 6
 7
 8
 9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;server&quot;</span><span class="p">:</span> <span class="s2">&quot;https://platform.genexplain.com&quot;</span><span class="p">,</span>
    <span class="nt">&quot;password&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="nt">&quot;reconnect&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
    <span class="nt">&quot;user&quot;</span><span class="p">:</span> <span class="s2">&quot;&quot;</span><span class="p">,</span>
    <span class="nt">&quot;tasks&quot;</span><span class="p">:</span> <span class="p">[{</span>
        <span class="nt">&quot;wait&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
        <span class="nt">&quot;method&quot;</span><span class="p">:</span> <span class="s2">&quot;Regulator search&quot;</span><span class="p">,</span>
        <span class="nt">&quot;workflow&quot;</span><span class="p">:</span> <span class="kc">false</span><span class="p">,</span>
        <span class="nt">&quot;progress&quot;</span><span class="p">:</span> <span class="kc">true</span><span class="p">,</span>
        <span class="nt">&quot;do&quot;</span><span class="p">:</span> <span class="s2">&quot;analyze&quot;</span><span class="p">,</span>
        <span class="nt">&quot;parameters&quot;</span><span class="p">:</span> <span class="p">[</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;sourcePath&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2639, Affymetrix HG-U133A microarray/Data/DEGs with limma/Normalized (RMA) DEGs with limma/Condition_1 vs. Condition_2/Enriched motifs Summary filtered TFs Genes Entrez&quot;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;weightColumn&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;Avg. adj. site FE&quot;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;isInputSizeLimited&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="kc">false</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;inputSizeLimit&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mi">100000</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;maxRadius&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mi">3</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;scoreCutoff&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mf">0.1</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;bioHub&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;GeneWays hub&quot;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;species&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;Human (Homo sapiens)&quot;</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;calculatingFDR&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="kc">true</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;FDRcutoff&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mf">0.01</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;ZScoreCutoff&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mf">1.1</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;penalty&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mf">0.5</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;decorators&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="p">[[</span>
                    <span class="p">{</span>
                        <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;decoratorName&quot;</span><span class="p">,</span>
                        <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;Apply Context&quot;</span>
                    <span class="p">},</span>
                    <span class="p">{</span>
                        <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;parameters&quot;</span><span class="p">,</span>
                        <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="p">[</span>
                            <span class="p">{</span>
                                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;tableName&quot;</span><span class="p">,</span>
                                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Examples/TNF-stimulation of HUVECs GSE2639, Affymetrix HG-U133A microarray/Data/DEGs with limma/Treatment vs. Control Genes Entrez&quot;</span>
                            <span class="p">},</span>
                            <span class="p">{</span>
                                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;tableColumn&quot;</span><span class="p">,</span>
                                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;logFC&quot;</span>
                            <span class="p">},</span>
                            <span class="p">{</span>
                                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;decayFactor&quot;</span><span class="p">,</span>
                                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="mf">0.01</span>
                            <span class="p">}</span>
                        <span class="p">]</span>
                    <span class="p">}</span>
                <span class="p">]]</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;isoformFactor&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="kc">true</span>
            <span class="p">},</span>
            <span class="p">{</span>
                <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;outputTable&quot;</span><span class="p">,</span>
                <span class="nt">&quot;value&quot;</span><span class="p">:</span> <span class="s2">&quot;data/Projects/Demo project/example_regulator_search&quot;</span>
            <span class="p">}</span>
        <span class="p">]</span>
    <span class="p">}],</span>
    <span class="nt">&quot;verbose&quot;</span><span class="p">:</span> <span class="kc">true</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h2 id="exec-executing-tasks-using-the-json-interface">exec - Executing tasks using the JSON interface</h2>
<p>The <em>exec</em> application provides a rich interface to interact with platform servers using JSON documents to configure tasks. It is possible to create complex workflows including re-usable templates, loops and conditional branch points. The <a href="../json_hello_world/">Hello world-tutorial</a> demonstrates several ways of how to make use of this interface.
A JSON document is an object in which the property <em>do</em> specifies an executor. An executor is a function provided by the interface, e.g. to import or export data. Parameters of the executor are also specified as properties of the JSON object.</p>
<h3 id="main-executors">Main executors</h3>
<p>The following table describes available executors. Most of them correspond to public methods of the <a href="../java_api/">Java API</a>, but some are only provided by the JSON interface, e.g. <em>branch</em> or <em>external</em>. Each executor and its JSON configuration is described in more detail in the following sections. Their usage is also demonstrated by several examples.</p>
<table>
<thead>
<tr>
<th>Name</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><strong>analyze</strong></td>
<td>Calls the analysis method with specified parameters</td>
</tr>
<tr>
<td><strong>branch</strong></td>
<td>Executes a branch point</td>
</tr>
<tr>
<td><strong>createFolder</strong></td>
<td>Creates a folder</td>
</tr>
<tr>
<td><strong>export</strong></td>
<td>Export data using a dedicated exporter</td>
</tr>
<tr>
<td><strong>external</strong></td>
<td>Runs an external tool</td>
</tr>
<tr>
<td><strong>get</strong></td>
<td>Gets specified table</td>
</tr>
<tr>
<td><strong>imPort</strong></td>
<td>Import a data file using a dedicated importer</td>
</tr>
<tr>
<td><strong>itemParameters</strong></td>
<td>Lists parameters for specified application, importer, or exporter</td>
</tr>
<tr>
<td><strong>jobStatus</strong></td>
<td>Gets the status for a job id</td>
</tr>
<tr>
<td><strong>listItems</strong></td>
<td>Lists available application, importer, or exporter items</td>
</tr>
<tr>
<td><strong>list</strong></td>
<td>Lists contents of specified folder</td>
</tr>
<tr>
<td><strong>put</strong></td>
<td>Puts table into specified folder</td>
</tr>
<tr>
<td><strong>setParameters</strong></td>
<td>Sets/adds/removes parameter strings</td>
</tr>
</tbody>
</table>
<h3 id="analyze">analyze</h3>
<p>The <em>analyze</em> executor runs an analysis tool or workflow on the remote server.</p>
<ul>
<li><strong>method</strong> - Name of the platform tool or workflow, where orkflows are specified by their platform path.</li>
<li><strong>parameters</strong> - A JSON object with parameters of the analysis tool or workflow.</li>
<li><strong>workflow</strong> - Set <code>true</code> if the called method is a workflow.</li>
<li><strong>wait</strong> - Set <code>true</code> to wait for the analysis to finish.</li>
<li><strong>progress</strong>  - Set <code>true</code> to obtain more progress information.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>         <span class="s2">&quot;analyze&quot;</span><span class="p">,</span>
    <span class="nt">&quot;method&quot;</span><span class="p">:</span>     <span class="s2">&quot;name of tool or path to workflow&quot;</span><span class="p">,</span>
    <span class="nt">&quot;parameters&quot;</span><span class="p">:</span> <span class="p">{</span> <span class="p">},</span>
    <span class="nt">&quot;workflow&quot;</span><span class="p">:</span>   <span class="kc">false</span><span class="p">,</span>
    <span class="nt">&quot;wait&quot;</span><span class="p">:</span>       <span class="kc">false</span><span class="p">,</span>
    <span class="nt">&quot;progress&quot;</span><span class="p">:</span>   <span class="kc">false</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="branch">branch</h3>
<p>Selects the next task or task set using a branch selector. </p>
<ul>
<li><strong>branchSelector</strong> - The canonical name of the Java class that implements the executor.</li>
</ul>
<p>The JSON document can contain further properties that configure the selector. In addition, the JSON configuration that was used to invoke the <em>exec</em> application is handed to the selector. Please see example selector implementations in this source repository.</p>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>             <span class="s2">&quot;branch&quot;</span><span class="p">,</span>
    <span class="nt">&quot;branchSelector&quot;</span><span class="p">:</span> <span class="s2">&quot;com.branch.selector.Class&quot;</span><span class="p">,</span>
    <span class="nt">&quot;other parameters&quot;</span><span class="p">:</span> <span class="s2">&quot;further properties used by the selector&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="createfolder">createFolder</h3>
<ul>
<li><strong>path</strong> - Path of the parent folder.</li>
<li><strong>name</strong> - Name of the new folder.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>   <span class="s2">&quot;createFolder&quot;</span><span class="p">,</span>
    <span class="nt">&quot;path&quot;</span><span class="p">:</span> <span class="s2">&quot;platform path&quot;</span><span class="p">,</span>
    <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;name of new folder&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="export">export</h3>
<p>Exports an item from the platform workspace to a local file.</p>
<ul>
<li><strong>file</strong> - Local file to create for the export.</li>
<li><strong>path</strong> - Path of the platform item to export.</li>
<li><strong>exporter</strong> - Name of the exporter to apply, e.g. <code>Tab-separated text (*.txt)</code> for a text table.</li>
<li><strong>parameters</strong> - Parameters to be specified to the exporter.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>         <span class="s2">&quot;export&quot;</span><span class="p">,</span>
    <span class="nt">&quot;file&quot;</span><span class="p">:</span>       <span class="s2">&quot;local file path to store export&quot;</span><span class="p">,</span>
    <span class="nt">&quot;path&quot;</span><span class="p">:</span>       <span class="s2">&quot;platform path to export&quot;</span><span class="p">,</span>
    <span class="nt">&quot;exporter&quot;</span><span class="p">:</span>   <span class="s2">&quot;name of exporter&quot;</span><span class="p">,</span>
    <span class="nt">&quot;parameters&quot;</span><span class="p">:</span> <span class="s2">&quot;JsonObject with exporter parameters&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="external">external</h3>
<p>This executor invokes an external program, e.g. a C++ application or an R script.</p>
<ul>
<li><strong>bin</strong> - The command to be executed.</li>
<li><strong>params</strong> - List of parameters to be specified to the external program.</li>
<li><strong>showOutput</strong> - Set <code>true</code> to get output of the external program printed to standard output.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>         <span class="s2">&quot;external&quot;</span><span class="p">,</span>
    <span class="nt">&quot;bin&quot;</span><span class="p">:</span>        <span class="s2">&quot;command to execute&quot;</span><span class="p">,</span>
    <span class="nt">&quot;params&quot;</span><span class="p">:</span>     <span class="s2">&quot;simple string or array of parameters to add to commandline&quot;</span><span class="p">,</span>
    <span class="nt">&quot;showOutput&quot;</span><span class="p">:</span> <span class="s2">&quot;set true to observe standard output of invoked tool&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="get">get</h3>
<p>Downloads a table from the platform workspace and optionally stores it in a local file and/or prints it to standard output.</p>
<ul>
<li><strong>table</strong> - Platform path of the table to download.</li>
<li><strong>toFile</strong> - Path of file to which to write table.</li>
<li><strong>toStdout</strong> - Set <code>true</code> to get table printed on standard output.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>    <span class="s2">&quot;get&quot;</span><span class="p">,</span>
    <span class="nt">&quot;table&quot;</span><span class="p">:</span> <span class="s2">&quot;platform path of table to download&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toFile&quot;</span><span class="p">:</span> <span class="s2">&quot;path of local file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toStdout&quot;</span><span class="p">:</span> <span class="s2">&quot;to print table object to standard output&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="import">imPort</h3>
<p>Imports a local file to the platform workspace.</p>
<ul>
<li><strong>file</strong> - Local file to import.</li>
<li><strong>path</strong> - Platform path including the name of imported item.</li>
<li><strong>importer</strong> - The importer to apply, e.g. <code>Tabular (*.txt, *.xls, *.tab, etc.)</code> for tables.</li>
<li><strong>parameters</strong> - Parameters for the importer.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>         <span class="s2">&quot;imPort&quot;</span><span class="p">,</span>
    <span class="nt">&quot;file&quot;</span><span class="p">:</span>       <span class="s2">&quot;local file to import&quot;</span><span class="p">,</span>
    <span class="nt">&quot;path&quot;</span><span class="p">:</span>       <span class="s2">&quot;the designated import location in the platform&quot;</span><span class="p">,</span>
    <span class="nt">&quot;importer&quot;</span><span class="p">:</span>   <span class="s2">&quot;name of importer&quot;</span><span class="p">,</span>
    <span class="nt">&quot;parameters&quot;</span><span class="p">:</span> <span class="s2">&quot;JsonObject with importer parameters&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="itemparameters">itemParameters</h3>
<p>This is a collectiv function that can get information about the available parameters for an analysis tool (application), exporter or importer. If the <em>item</em> is an exporter or importer, one needs to specify the corresponding target platform path to export or import.</p>
<ul>
<li><strong>name</strong> - Item name which may be the name of an analysis tool, an exporter or an importer name.</li>
<li><strong>type</strong> - Type of item, one of application, exporter or importer.</li>
<li><strong>path</strong> - If the item is an exporter or importer, the corresponding platform path is required to determine possible context-dependent parameters.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>   <span class="s2">&quot;itemParameters&quot;</span><span class="p">,</span>
    <span class="nt">&quot;name&quot;</span><span class="p">:</span> <span class="s2">&quot;name of item for which to get parameters&quot;</span><span class="p">,</span>
    <span class="nt">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;type of item for which to get parameters: application, exporter, importer&quot;</span><span class="p">,</span>
    <span class="nt">&quot;path&quot;</span><span class="p">:</span> <span class="s2">&quot;if exporter or importer, path for which to get parameters in context with ex-/importer&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="jobstatus">jobStatus</h3>
<p>Returns the status of a running analysis job.</p>
<ul>
<li><strong>jobId</strong> - Id of the job whose status is requested.</li>
<li><strong>toFile</strong> - Path of file to which to write output.</li>
<li><strong>toStdout</strong> - Set <code>true</code> to get output printed on standard output.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>     <span class="s2">&quot;jobStatus&quot;</span><span class="p">,</span>
    <span class="nt">&quot;jobId&quot;</span><span class="p">:</span>  <span class="s2">&quot;id of platform job to request status&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toFile&quot;</span><span class="p">:</span> <span class="s2">&quot;path of local file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toStdout&quot;</span><span class="p">:</span> <span class="s2">&quot;to print output to standard output&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="listitems">listItems</h3>
<p>Gets listings of available applications, importers or exporters.</p>
<ul>
<li><strong>type</strong> - Type of items to list, one of application, exporter or importer.</li>
<li><strong>toFile</strong> - Path of file to which to write output.</li>
<li><strong>toStdout</strong> - Set <code>true</code> to get output printed on standard output.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>   <span class="s2">&quot;listItems&quot;</span><span class="p">,</span>
    <span class="nt">&quot;type&quot;</span><span class="p">:</span> <span class="s2">&quot;type of items to list: applications, importers, or exporters&quot;</span>
    <span class="s2">&quot;toFile&quot;</span><span class="p">:</span> <span class="s2">&quot;path of local file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toStdout&quot;</span><span class="p">:</span> <span class="s2">&quot;to print output to standard output&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="list">list</h3>
<p>Gets the listing of specified folder.</p>
<ul>
<li><strong>folder</strong> - Platform folder to get listing for.</li>
<li><strong>toFile</strong> - Path of file to which to write output.</li>
<li><strong>toStdout</strong> - Set <code>true</code> to get output printed on standard output.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>     <span class="s2">&quot;list&quot;</span><span class="p">,</span>
    <span class="nt">&quot;folder&quot;</span><span class="p">:</span> <span class="s2">&quot;platform folder to get listing for&quot;</span>
    <span class="s2">&quot;toFile&quot;</span><span class="p">:</span> <span class="s2">&quot;path of local file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;toStdout&quot;</span><span class="p">:</span> <span class="s2">&quot;to print output to standard output&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="put">put</h3>
<p>Uploads a table from specified local file into the platform workspace.</p>
<ul>
<li><strong><em>file</em></strong> - Local file with data table to upload.</li>
<li><strong><em>skip</em></strong> - Number of lines to skip in the beginning of the file.</li>
<li><strong><em>delimiter</em></strong> - Column delimiter string.</li>
<li><strong><em>table</em></strong> - JSON array of arrays with data columns. Note that the table has no title row, but columns are specified separately.</li>
<li><strong><em>columns</em></strong> - Column definition by an array of two-element arrays giving column name and type, where type is one of <code>Integer, Float, Boolean, or Text</code>.</li>
<li><strong><em>path</em></strong> - Platform path to put table which includes the name of the table item in the platform workspace.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7
8
9</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>        <span class="s2">&quot;put&quot;</span><span class="p">,</span>
    <span class="nt">&quot;file&quot;</span><span class="p">:</span>      <span class="s2">&quot;file with data table to put into platform&quot;</span><span class="p">,</span>
    <span class="nt">&quot;skip&quot;</span><span class="p">:</span>      <span class="s2">&quot;number of lines to skip in the beginning of input file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;delimiter&quot;</span><span class="p">:</span> <span class="s2">&quot;delimiter of table columns in input file&quot;</span><span class="p">,</span>
    <span class="nt">&quot;table&quot;</span><span class="p">:</span>     <span class="s2">&quot;array of arrays with data to put into platform&quot;</span><span class="p">,</span>
    <span class="nt">&quot;columns&quot;</span><span class="p">:</span>   <span class="s2">&quot;array of two-element arrays specifying column names and type. The latter can be Integer, Float, Boolean, Text&quot;</span><span class="p">,</span>
    <span class="nt">&quot;path&quot;</span><span class="p">:</span>      <span class="s2">&quot;destination path of table in platform&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>

<h3 id="setparameters">setParameters</h3>
<p>Sets/adds/removes string replacements in the parameter object that will modify parameters of subsequent tasks.</p>
<ul>
<li><strong><em>set</em></strong> - a JSON object whose keys can be found in the <em>replaceStrings</em> array of the parameter object and whose value will replace the current one.</li>
<li><strong><em>remove</em></strong> - a JSON object whose keys will be removed from the <em>replaceStrings</em> array of the parameter object.</li>
<li><strong><em>before</em></strong> - an array of two-element arrays that are inserted in the beginning of the <em>replaceStrings</em> array.</li>
<li><strong><em>after</em></strong> - an array of two-element arrays that are appended to the <em>replaceStrings</em> array.</li>
</ul>
<table class="codehilitetable"><tr><td class="linenos"><div class="linenodiv"><pre>1
2
3
4
5
6
7</pre></div></td><td class="code"><div class="codehilite"><pre><span></span><span class="p">{</span>
    <span class="nt">&quot;do&quot;</span><span class="p">:</span>     <span class="s2">&quot;setParameters&quot;</span><span class="p">,</span>
    <span class="nt">&quot;set&quot;</span><span class="p">:</span>    <span class="s2">&quot;sets existing parameter to specified value&quot;</span><span class="p">,</span>
    <span class="nt">&quot;remove&quot;</span><span class="p">:</span> <span class="s2">&quot;removes existing parameter&quot;</span><span class="p">,</span>
    <span class="nt">&quot;before&quot;</span><span class="p">:</span> <span class="s2">&quot;add parameter before others in the parameter array&quot;</span><span class="p">,</span>
    <span class="nt">&quot;after&quot;</span><span class="p">:</span>  <span class="s2">&quot;add parameter at the end of the parameter array&quot;</span>
<span class="p">}</span>
</pre></div>
</td></tr></table>
                
                  
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
          <a href="../java_json_manual_intro/" title="Introduction" class="md-flex md-footer-nav__link md-footer-nav__link--prev" rel="prev">
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-back md-footer-nav__button"></i>
            </div>
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Previous
                </span>
                Introduction
              </span>
            </div>
          </a>
        
        
          <a href="../json_hello_world/" title="JSON interface Hello world tutorial" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                JSON interface Hello world tutorial
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
          <div class="md-footer-copyright__highlight">
            &copy; 2019 geneXplain GmbH, Wolfenbuettel, Germany
          </div>
        
        powered by
        <a href="https://www.mkdocs.org">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/">
          Material for MkDocs</a>
      </div>
      
        
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="../assets/javascripts/application.5e60981f.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:".."}})</script>
      
        <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
      
    
    
      
    
  </body>
</html>